flink-dstl/flink-dstl-dfs/src/main/java/org/apache/flink/changelog/fs/ChangelogStreamHandleReaderWithCache.java

```diff
@@ -0,0 +1,203 @@
+        return cache.compute(
+                fileHandle.getFilePath(),
+                (key, oldValue) -> {
+                    if (oldValue == null) {
+                        oldValue = downloadToCacheFile(fileHandle);
+                    }
+                    oldValue.retain();
+                    return oldValue;
+                });
+    }
+
+    private RefCountedFile downloadToCacheFile(FileStateHandle fileHandle) {
+        RefCountedFile refCountedFile = null;
+
+        try (FSDataInputStream inputStream = fileHandle.openInputStream()) {
+            File directory = cacheDirectories[next.getAndIncrement() % cacheDirectories.length];
+            File file = File.createTempFile(CACHE_FILE_PREFIX, null, directory);
+
+            IOUtils.copyBytes(wrap(inputStream), new FileOutputStream(file));
+            LOG.debug(
+                    "download and decompress dstl file : {} to cache file : {}",
+                    fileHandle.getFilePath(),
+                    refCountedFile.getFile().getPath());
+
+            refCountedFile = new RefCountedFile(file);
+        } catch (IOException e) {
+            ExceptionUtils.rethrow(e);
+        }
+
+        return refCountedFile;
+    }
+
+    private FileInputStream openAndSeek(RefCountedFile refCountedFile, long offset)
+            throws IOException {
+        FileInputStream fin = new FileInputStream(refCountedFile.getFile());
+        if (offset != 0) {
+            LOG.debug("seek to {}", offset);
+            fin.getChannel().position(offset);
+        }
+        return fin;
+    }
+
+    private DataInputStream wrapStream(Path dfsPath, FileInputStream fin) {
+        return new DataInputStream(new BufferedInputStream(fin)) {
+            @Override
+            public void close() throws IOException {
+                try {
+                    super.close();
+                } finally {
+                    cache.computeIfPresent(
+                            dfsPath,
+                            (key, value) -> {
+                                value.release();
+                                if (value.getReferenceCounter() == NO_USING_REF_COUNT) {
+                                    cacheCleanScheduler.schedule(
+                                            () -> cleanCacheFile(dfsPath),
+                                            cacheIdleMillis,
+                                            TimeUnit.MILLISECONDS);
+                                }
+                                return value;
+                            });
+                }
+            }
+        };
+    }
+
+    private void cleanCacheFile(Path dfsPath) {
+        cache.computeIfPresent(
+                dfsPath,
+                (key, value) -> {
+                    if (value.getReferenceCounter() == NO_USING_REF_COUNT) {
+                        LOG.debug("clean cached file : {}", value.getFile().getPath());
+                        value.release();
+                        return null;
```

Comment:
This `null` means the mapping will be removed, right?

Reply:
Yes, may be we should add some annotations ?

Reply:
Good idea, a short comment would be helpful.